### ISSUE TITLE
com.google.gson.JsonParseException unexpectedly thrown by the API

### ISSUE THREAD

**Issue by sevokn** [cloudant/java-cloudant#311]
I'm using a `ViewRequest`s to query views (secondary indexes) on the server.

If while parsing the JSON response a network problem occurs and the socket that is used is reset, causing a `java.net.SocketException` to be thrown, a stack trace like the following is produced:

```
com.google.gson.JsonSyntaxException: java.net.SocketException: Connection reset
        at com.google.gson.Gson.fromJson(Gson.java:902) ~[gson-2.7.jar:?]
        at com.google.gson.Gson.fromJson(Gson.java:825) ~[gson-2.7.jar:?]
        at com.cloudant.client.org.lightcouch.internal.CouchDbUtil.getResponse(CouchDbUtil.java:229) ~[cloudant-client-2.6.2.jar:?]
        at com.cloudant.client.internal.views.ViewRequester.executeRequestWithResponseAsJson(ViewRequester.java:35) ~[cloudant-client-2.6.2.jar:?]
        at com.cloudant.client.internal.views.ViewRequester.getResponseAsJson(ViewRequester.java:28) ~[cloudant-client-2.6.2.jar:?]
        at com.cloudant.client.internal.views.ViewRequestImpl.getResponse(ViewRequestImpl.java:34) ~[cloudant-client-2.6.2.jar:?]
        [my classes calling getResponse() removed]
Caused by: java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:196) ~[?:1.7.0_80]
        at java.net.SocketInputStream.read(SocketInputStream.java:122) ~[?:1.7.0_80]
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:235) ~[?:1.7.0_80]
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:275) ~[?:1.7.0_80]
        at java.io.BufferedInputStream.read(BufferedInputStream.java:334) ~[?:1.7.0_80]
        at sun.net.www.http.ChunkedInputStream.fastRead(ChunkedInputStream.java:244) ~[?:1.7.0_80]
        at sun.net.www.http.ChunkedInputStream.read(ChunkedInputStream.java:689) ~[?:1.7.0_80]
        at java.io.FilterInputStream.read(FilterInputStream.java:133) ~[?:1.7.0_80]
        at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3066) ~[?:1.7.0_80]
        at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:283) ~[?:1.7.0_80]
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:325) ~[?:1.7.0_80]
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:177) ~[?:1.7.0_80]
        at java.io.InputStreamReader.read(InputStreamReader.java:184) ~[?:1.7.0_80]
        at com.google.gson.stream.JsonReader.fillBuffer(JsonReader.java:1287) ~[gson-2.7.jar:?]
        at com.google.gson.stream.JsonReader.nextQuotedValue(JsonReader.java:1016) ~[gson-2.7.jar:?]
        at com.google.gson.stream.JsonReader.nextString(JsonReader.java:815) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:718) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:739) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:739) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:739) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:731) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:739) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:714) ~[gson-2.7.jar:?]
        at com.google.gson.internal.bind.TypeAdapters$35$1.read(TypeAdapters.java:910) ~[gson-2.7.jar:?]
        at com.google.gson.Gson.fromJson(Gson.java:887) ~[gson-2.7.jar:?]
        ... 14 more
```

This is a bit problematic, since `com.google.gson.JsonSyntaxException` extends `com.google.gson.JsonParseException` which in turn extends `java.lang.RuntimeException`, and is thus not a checked exception and consequently is not declared in the method signature for `ViewRequest.getResponse()` in the java-cloudant API. So in addition to `CouchDbException` (see #309),  `JsonParseException` (and its subclasses) is another unchecked exception that the user of the java-cloudant API needs to guard against when using the API.

Could you wrap the `JsonParseException` (and its subclasses) into a `CouchDbException`, in order to keep the API consistent? In a way you are leaking a Gson-specific exception to the user and it would be more clear and consistent to only have to deal with `CouchDbException`s.

**Comment by rhyshort** [comment]
Thanks for reporting this. We'll look into doing it as part of the next Major release. Exceptions for the API surface are definitely in need of improvement, and as it stands it's very much a legacy of light couch and something I want to have improved.

**Comment by ricellis** [comment]
The "next major" will be the new [cloudant-java-sdk](https://github.com/IBM/cloudant-java-sdk). This problem has been resolved there since [0.0.21](https://github.com/IBM/cloudant-java-sdk/releases/tag/v0.0.21) - exceptions from deserializing into built-in model classes there are wrapped in the base `ServiceResponseException`. The Cloudant API reference documentation has [more detail about error handling](https://cloud.ibm.com/apidocs/cloudant?code=java#code-example-java) in the new SDK and we recently published a [migration guide](https://github.com/cloudant/java-cloudant/blob/master/MIGRATION.md).
